storage/innobase/gis/gis0sea.cc

```diff
@@ -278,9 +278,8 @@ rtr_pcur_getnext_from_path(
 				      ulint{RW_S_LATCH}, "");
 			static_assert(ulint{BTR_MODIFY_LEAF} ==
 				      ulint{RW_X_LATCH}, "");
-			rw_latch = (my_latch_mode | 4) == BTR_CONT_MODIFY_TREE
-				? RW_NO_LATCH
-				: rw_lock_type_t(my_latch_mode);
+			rw_latch = rw_lock_type_t(my_latch_mode
+						  & (RW_X_LATCH | RW_S_LATCH));
 		} else {
 			rw_latch = RW_X_LATCH;
```

Comment:
A RW_X_LATCH latch is now being set on both sides of the condition. Is there a reason for the additional RW_S_LATCH in the if part?

Reply:
The `if` part is a bit convoluted, but it can actually assign `rw_latch` to either `RW_X_LATCH` or `RW_S_LATCH`. In 75096c84b44875b5d226a734fffb08578bc21e96 I redefined many `btr_latch_mode` constants in such a way that we can simply convert them to `rw_lock_type_t` or `mtr_memo_type_t` with simple bitmasking. The changed statement should be equivalent to the following, if we assume that the variables are integers:
```c++
rw_latch = my_latch_mode & 3;
```
For example, we convert `BTR_MODIFY_TREE` and `BTR_MODIFY_LEAF` to `RW_X_LATCH` and `BTR_SEARCH_TREE` and `BTR_SEARCH_LEAF` to `RW_S_LATCH`.

Conditional branches can be very expensive on modern superscalar processors. In [MDEV-30567](https://jira.mariadb.org/browse/MDEV-30567) and 66b21ed5403aa4ef18a4c797af353ec3379b152d you can find another example where using bitwise arithmetics might make code harder to read for some, but faster to execute.

Reply:
Thank you for explaining the rationale for the change. It makes a lot more sense now :-)